jax/_src/lax/parallel.py

```diff
@@ -405,50 +405,94 @@ def pdot(x, y, axis_name, pos_contract=((), ()), pos_batch=((), ())):
                      pos_contract=pos_contract, pos_batch=pos_batch)
 
 
-def xeinsum(spec: str, x, y):
+def xeinsum(spec: str, *operands):
   in_spec, out_spec = spec.split('->')
-  (lhs_subs, lhs_named), (rhs_subs, rhs_named) = XeinsumSpecParser(in_spec).parse_args()
-  (out_subs, out_named), = XeinsumSpecParser(out_spec).parse_args()
-  all_named = {*lhs_named, *rhs_named, *out_named}
-  all_subs = {*lhs_subs, *rhs_subs, *out_subs}
-  lhs_uniques = set(lhs_subs) - set(rhs_subs)
-  rhs_uniques = set(rhs_subs) - set(lhs_subs)
-  if all_subs & all_named:
-    raise NotImplementedError
-  if not set(out_named).issubset({*lhs_named, *rhs_named}):
-    raise ValueError
-
-  # if a named axis appears in both inputs and not the output, contract!
-  named_contract = list(all_named - set(out_named))
-
-  # if a subscript appears in both inputs and not the outputs, contract!
-  subs_contract = all_subs - set(out_subs)
-
-  lhs_reduce_axes = [lhs_subs.index(n) for n in lhs_uniques & subs_contract]
-  if lhs_reduce_axes:
-    x = lax._reduce_sum(x, lhs_reduce_axes)
-    for i in sorted(lhs_reduce_axes, reverse=True):
-      del lhs_subs[i]
-
-  rhs_reduce_axes = [rhs_subs.index(n) for n in rhs_uniques & subs_contract]
-  if rhs_reduce_axes:
-    y = lax._reduce_sum(y, rhs_reduce_axes)
-    for i in sorted(rhs_reduce_axes, reverse=True):
-      del rhs_subs[i]
-
-  pos_contract = unzip2((lhs_subs.index(n), rhs_subs.index(n))
-                        for n in subs_contract - (lhs_uniques | rhs_uniques))
-
-  # if a subscript apperas in both inputs _and_ the outputs, batch!
-  subs_batch = all_subs - subs_contract
-  if subs_batch & (lhs_uniques | rhs_uniques):
-    raise NotImplementedError
-
-  pos_batch = unzip2((lhs_subs.index(n), rhs_subs.index(n))
-                        for n in subs_batch)
-
-  return pdot(x, y, axis_name=named_contract,
-              pos_contract=pos_contract, pos_batch=pos_batch)
+
+  if len(operands) == 1:
+    x = operands[0]
+    (in_subs, in_named), = XeinsumSpecParser(in_spec).parse_args()
+    (out_subs, out_named), = XeinsumSpecParser(out_spec).parse_args()
+
+    # output subs and named axes must appear in at least one of the inputs.
+    if not set(out_named).issubset(set(in_named)):
+      raise ValueError(f"Found named axes {set(out_named) - set(in_named)} "
+                       "appearing in the output spec but not in the input.")
+    if not set(out_subs).issubset(set(in_subs)):
+      raise ValueError(f"Found subscript(s) {set(out_subs) - set(in_subs)} "
+                       "appearing in the output spec but not in the input.")
+
+    # reduce positional axes.
+    subs_reduce = list(set(in_subs) - set(out_subs))
+    subs_reduce_axes = [in_subs.index(n) for n in subs_reduce]
+    x = lax._reduce_sum(x, subs_reduce_axes)
+
+    # reduce named axes.
+    named_reduce_axes = list(set(in_named) - set(out_named))
+    for ax in named_reduce_axes:
+      x = psum(x, ax)
+    return x
+  elif len(operands) == 2:
```

Comment:
Is this casing at the top level really necessary? It would be better to stop assuming that the output of `XeinsumSpecParser` has only one or two output but process it as a list instead. At the moment we will effectively have to review and maintain two parallel implementations that could be handled by a single algorithm.

Reply:
Yeah -- I think that is a good idea, let me take a look.

BTW is there any plan to use opt_einsum to find a contraction path like in regular einsum (defined in lax_numpy.py)? That way we'd only ever have to worry about unary or binary einsums. Or is it that depending on 'axis_resources' you'd need a different cost metric for the path contraction that perhaps opt_einsum cannot handle?

Reply:
Great question! In the future we definitely should be using opt_einsum. We might need to do a little munging on the einsum strings to get them to a form that's compatible with that library, but it should be doable. But there's no need to include it in this PR. For now we can encode any (possibly silly) strategy for n-ary contractions.

Reply:
Thanks! For this PR I am not adding n-ary contractions to keep the scope limited. (Maybe in a future PR)

This should be ready for another review. Please take another look!